### ISSUE TITLE
InnerAssignment reports violation when && condition is in WHILE expression

### ISSUE THREAD

**Issue by kec2** [checkstyle/checkstyle#6805]
https://checkstyle.org/config_coding.html#InnerAssignment

```
/var/tmp$ cat TestConfig.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
       <module name="InnerAssignment" />
    </module>
</module>

/var/tmp$ cat TestClass.java 
public class TestClass {
  void method() {
    while ((line = file.read()) != null && noOfSQLErrors < maxErrors) {} // violation
    while ((line = file.read()) != null) {} // no violatin; specially allowed by Check, see doc
  }
}

/var/tmp$ java -jar checkstyle-8.21-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] /var/tmp/TestClass.java:3:18: Inner assignments should be avoided. [InnerAssignment]
Audit done.
Checkstyle ends with 1 errors.
```

Expected: no violations

**Comment by rnveach** [comment]
@kec2 Please look over http://checkstyle.sourceforge.net/report_issue.html#How_to_report_a_bug.3F and http://checkstyle.sourceforge.net/report_issue.html#How_to_request_new_feature_for_existing_functionality.3F .
Be more specific with your request. "flexible/loose" is too vague to understand what you mean.
Also I don't see how the problem is the `<` condition when it is the other condition that has the assignment.

**Comment by kec2** [comment]
This example results in zero checkstyle errors:
```
String sqlFilePath;
while ((inputLine = sqlFile.readLine()) != null) {
...
}
```

This example results in one checkstyle error:

```
int maxErrors = 4;
int noOfSQLErrors  = 0;
String sqlFilePath;
while ((inputLine = sqlFile.readLine()) != null && noOfSQLErrors < maxErrors) {
```

The reason is the second part of the condition:
´´´
&& noOfSQLErrors < maxErrors
´´´

What I mean with more loose is that the InnerAssignment check should not fail because of conditions that has nothing to due with the assignment.

**Comment by rnveach** [comment]
@kec2 Why don't you use an xpath suppression similar to the following:
````
$ cat TestClass.java
public class TestClass {
    void method() {
int maxErrors = 4;
int noOfSQLErrors  = 0;
String sqlFilePath;

while ((inputLine = sqlFile.readLine()) != null && noOfSQLErrors < maxErrors) {}
while ((inputLine = sqlFile.readLine()) != null) {}
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="InnerAssignmentCheck" />
<module name="SuppressionXpathSingleFilter">
     <property name="checks" value="InnerAssignmentCheck" />
     <property name="query" value="//LITERAL_WHILE//LAND//ASSIGN"/>
</module>
    </module>
</module>

$ java -jar checkstyle-8.21-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

I don't know xpath that well, but you should be able to specify for the while statement that has an `&&` or `||` in it with the assignment.

**Comment by rnveach** [comment]
Fix was merged
